# Highlighting Differences in Data Visualizations

> This notebook serves the purpose of exploring Vega-Lite datasets and experimenting with them. Ultimately, it serves as a backbone to a bachelor thesis with the same topic name.

In [1]:
import os
import json

import pandas as pd
import altair as alt
from vega import VegaLite

### Offline Access to Vega Datasets

In [2]:
# pip install vega_datasets
from vega_datasets import data
vega_datasets = data

In [3]:
datasets_ = data.list_datasets()
print("List of all datasets:\n\n", datasets_)

List of all datasets:

 ['7zip', 'airports', 'annual-precip', 'anscombe', 'barley', 'birdstrikes', 'budget', 'budgets', 'burtin', 'cars', 'climate', 'co2-concentration', 'countries', 'crimea', 'disasters', 'driving', 'earthquakes', 'ffox', 'flare', 'flare-dependencies', 'flights-10k', 'flights-200k', 'flights-20k', 'flights-2k', 'flights-3m', 'flights-5k', 'flights-airport', 'gapminder', 'gapminder-health-income', 'gimp', 'github', 'graticule', 'income', 'iowa-electricity', 'iris', 'jobs', 'la-riots', 'londonBoroughs', 'londonCentroids', 'londonTubeLines', 'lookup_groups', 'lookup_people', 'miserables', 'monarchs', 'movies', 'normal-2d', 'obesity', 'ohlc', 'points', 'population', 'population_engineers_hurricanes', 'seattle-temps', 'seattle-weather', 'sf-temps', 'sp500', 'stocks', 'udistrict', 'unemployment', 'unemployment-across-industries', 'uniform-2d', 'us-10m', 'us-employment', 'us-state-capitals', 'volcano', 'weather', 'weball26', 'wheat', 'windvectors', 'world-110m', 'zipcodes']


In [4]:
# Only the .json datasets are of interest to us, therefore we filter out datasets of different file types.

datasets = []
for dataset in datasets_:
    try:
        path = getattr(data, dataset).filepath
        if path.endswith(".json"):
            datasets.append(dataset)
    except (AttributeError, ValueError):
        continue
        
datasets.remove('anscombe')
datasets.remove('ohlc')
print("Available datasets:", datasets)

Available datasets: ['barley', 'burtin', 'cars', 'crimea', 'driving', 'iris', 'wheat']


Because some datasets were not suitable (anscombe and ohlc), all remaining datasets with file extension .json have been copied to a folder called "datasets".

## randomize.py

Helper functions:<br>
```getNotXYColor(file)```: returns a list of those values that are NOT encoded in the x or y field of the vega lite spec.<br>
```modify(point, file)```: carries two actions: one to skew x and y values and one to add a new point based on statistical information about the dataset.<br><br>
Main function:<br>
```randomize(json_file, p)```: generates a new json file in altered directory. scans through each datapoint of the file. with probability p, the datapoint gets altered, meaning either removed (1/3) or sent to modify (2/3).

## compare.py

Main function:<br>
```compare(a_json, b_json)```: if spec is barplot, we iterate through datapoints in a and b. regardless of the datapoints y values, if they are identical, copy to new file and adjust y value.

# Datasets info

In [5]:
vega_datasets.wheat.description

"In an 1822 letter to Parliament, William Playfair[1]_, a Scottish engineer who is often credited as the founder of statistical graphics, published an elegant chart on the price of wheat[2]_. It plots 250 years of prices alongside weekly wages and the reigning monarch. He intended to demonstrate that 'never at any former period was wheat so cheap, in proportion to mechanical labour, as it is at the present time.' The electronic dataset was created by Mike Bostock and released into the public domain."

# The Datasets

> The following Altair Visualizations offer an option "View Source". <br>
> The source codes are manually saved.

In [6]:
os.chdir("datasets/datasets_altered")

### set percent for modification plots:

In [7]:
percent = 20

## (anscombe.json)

Anscombe’s Quartet is a famous dataset constructed by Francis Anscombe. The common summary statistics in each of the series identical, despite the subsets' different characteristics. <br> It makes sense to make subplots. Variations of this dataset do not make much sense, however for demonstration purposes in this project, we will generate them anyway.

In [8]:
# source = vega_datasets.anscombe()

# x = "X"
# y = "Y"
# color = "Series"

In [9]:
# alt.Chart(source).mark_point().encode(
#     x=x,
#     y=y,
#     color=color,
#     tooltip=(x,y)
# ).interactive().facet("Series", columns=2)

In [10]:
# alternation = None
# with open(f"anscombe{percent}.json", 'r') as f:
#     data = json.load(f)
#     alternation = pd.DataFrame(data)

In [11]:
# alt.Chart(alternation).mark_point().encode(
#     x=x,
#     y=y,
#     color=color,
#     tooltip=(x,y)
# ).interactive().facet("Series", columns=2)

## barley.json

The Becker’s Barley Trellis charts identify an anomoly in a widely used agriculatural dataset, which is called “The Morris Mistake”. It shows that "Morris" is the only site that is the reverse of other panels. It is usually displayed split into the sites, however I will aggregate the sites.

In [12]:
source = vega_datasets.barley()

alt.Chart(source).mark_point().encode(
    x = "yield",
    y = "site",
    color = "year:N",
    tooltip=("yield", "year")
)

alt.Chart(...)

In [13]:
alternation = None
with open(f"barley{percent}.json", 'r') as f:
    data = json.load(f)
    alternation = pd.DataFrame(data)

In [14]:
alt.Chart(alternation).mark_point().encode(
    x = "yield",
    y = "site",
    color = "year:N",
    tooltip=("yield", "year")
)

alt.Chart(...)

## burtin.json

This dataset was gathered by Will Burtin and is used to explore the effectiveness of various antibiotics in treating a variety of bacterial infections.

In [15]:
source = vega_datasets.burtin()

alt.Chart(source).mark_bar().encode(
    x = "Bacteria",
    y = alt.Y("Penicillin", scale=alt.Scale(type='pow', exponent=0.25)),
    color=alt.Color("Gram_Staining"),
    tooltip=[alt.Tooltip("Penicillin"), alt.Tooltip("Gram_Staining")]
)

alt.Chart(...)

In [16]:
alternation = None
with open(f"burtin{percent}.json", 'r') as f:
    data = json.load(f)
    alternation = pd.DataFrame(data)

In [17]:
alt.Chart(alternation).mark_bar().encode(
    x = "Bacteria",
    y = alt.Y("Penicillin", scale=alt.Scale(type='pow', exponent=0.25)),
    color=alt.Color("Gram_Staining"),
    tooltip=[alt.Tooltip("Penicillin"), alt.Tooltip("Gram_Staining")]
)

alt.Chart(...)

## cars.json [id]

Acceleration, horsepower, fuel efficiency, weight, and other characteristics of different makes and models of cars. 

In [18]:
id_cars = None

with open("../cars.json", "r") as f:
    data = json.load(f)
    id_cars = pd.DataFrame(data)

In [19]:
alt.Chart(id_cars).mark_point().encode(
    x = "Horsepower",
    y = "Miles_per_Gallon",
    color = "Origin",
    tooltip=("Name", "_ID_")
).interactive()

alt.Chart(...)

In [20]:
id_cars = None

with open(f"cars{percent}.json", "r") as f:
    data = json.load(f)
    id_cars = pd.DataFrame(data)

In [21]:
alt.Chart(id_cars).mark_point().encode(
    x = "Horsepower",
    y = "Miles_per_Gallon",
    color = "Origin",
    tooltip=("Name", "_ID_")
).interactive()

alt.Chart(...)

## crimea.json

This is a dataset containing monthly casualty counts from the Crimean war. 

In [22]:
source = None
with open(f"../crimea.json", 'r') as f:
    data = json.load(f)
    source = pd.DataFrame(data)
    
alt.Chart(source).mark_bar().encode(
    x = "date:N",
    y = "disease",
    color = "wounds",
    tooltip=("wounds", "disease", "other")
)

alt.Chart(...)

In [23]:
alternation = None
with open(f"crimea{percent}.json", 'r') as f:
    data = json.load(f)
    alternation = pd.DataFrame(data)

In [24]:
alt.Chart(source).mark_bar().encode(
    x = "date:N",
    y = "disease",
    color = "wounds",
    tooltip=("wounds", "disease", "other")
)

alt.Chart(...)

## driving.json [id]

This dataset tracks miles driven per capita along with gas prices annually from 1956 to 2010.

In [25]:
id_driving = None

with open("../driving.json", "r") as f:
    data = json.load(f)
    id_iris = pd.DataFrame(data)

In [26]:
alt.Chart(id_iris).mark_point().encode(
    x = "miles",
    y = "gas",
    color = "year",
    tooltip=("year", "miles", "_ID_")
).interactive()

alt.Chart(...)

In [27]:
id_driving = None

with open(f"driving{percent}.json", "r") as f:
    data = json.load(f)
    id_iris = pd.DataFrame(data)

In [28]:
alt.Chart(id_iris).mark_point().encode(
    x = "miles",
    y = "gas",
    color = "year",
    tooltip=("year", "miles", "_ID_")
).interactive()

alt.Chart(...)

## iris.json

This classic dataset contains lengths and widths of petals and sepals for 150 iris flowers, drawn from three species. It was introduced by R.A. Fisher in 1936.

In [29]:
x = "sepalLength"
y = "sepalWidth"

In [30]:
source = vega_datasets.iris()

alt.Chart(source).mark_point().encode(
    x=x,
    y=y,
    color='species',
    tooltip=(x,y)
).interactive()

alt.Chart(...)

In [31]:
# modified iris dataset available 5, 10, 15, 20

iris_modified = None
with open(f"iris{percent}.json", 'r') as f:
    data = json.load(f)
    iris_modified = pd.DataFrame(data)

In [32]:
iris_mod_plot = alt.Chart(iris_modified).mark_point().encode(
    x=x,
    y=y,
    color='species',
    tooltip=(x,y)
).interactive()

iris_mod_plot

alt.Chart(...)

## (ohlc.json)

(open, high, low and closed prices) - This one contains the performance of the Chicago Board Options Exchange

In [33]:
# source = vega_datasets.ohlc()

# alt.Chart(source).mark_bar().encode(
#     x='date:T',
#     color=alt.condition('datum.open <= datum.close',
#                         alt.value('#1a9850'), alt.value('#d73027')),
#     y='low:Q',
#     y2='high:Q',
#     tooltip=('date:T', 'open:Q', 'high:Q', 'low:Q', 'close:Q')
# )

In [34]:
# alternation = None
# with open(f"ohlc{percent}.json", 'r') as f:
#     data = json.load(f)
#     alternation = pd.DataFrame(data)

In [35]:
# alt.Chart(alternation).mark_bar().encode(
#     x='date:T',
#     color=alt.condition('datum.open <= datum.close',
#                         alt.value('#1a9850'), alt.value('#d73027')),
#     y='low:Q',
#     y2='high:Q',
#     tooltip=('date:T', 'open:Q', 'high:Q', 'low:Q', 'close:Q')
# )

## wheat.json

In an 1822 letter to Parliament, William Playfair, a Scottish engineer who is often credited as the founder of statistical graphics, published an elegant chart on the price of wheat. It plots 250 years of prices alongside weekly wages and the reigning monarch. He intended to demonstrate that 'never at any former period was wheat so cheap, in proportion to mechanical labour, as it is at the present time.' The electronic dataset was created by Mike Bostock and released into the public domain.

In [36]:
source = vega_datasets.wheat()

alt.Chart(source).mark_point().encode(
    x = "wages",
    y = "wheat",
    color = "year:Q",
    tooltip=("wages", "wheat")
).interactive()

alt.Chart(...)

In [37]:
alternation = None
with open(f"wheat{percent}.json", 'r') as f:
    data = json.load(f)
    alternation = pd.DataFrame(data)

In [38]:
alt.Chart(alternation).mark_point().encode(
    x = "wages",
    y = "wheat",
    color = "year:Q",
    tooltip=("wages", "wheat")
).interactive()

alt.Chart(...)

# Diverging Bar Plots

In [39]:
# example dataframe
data = pd.DataFrame({
    'Category': ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I'],
    'Value': [10, 8, 6, 4, 2, -2, -4, -6, -8]
})

plot = alt.Chart(data).mark_bar().encode(
    x='Category:N',
    y='Value:Q',
    color=alt.condition(
        alt.datum.Value >= 0,
        alt.value('steelblue'),  # Positive values
        alt.value('orange')  # Negative values
    )
)

# plot

# Example Uploads

In [40]:
os.chdir("../../fileUploadExamples")

In [41]:
example = None

with open("file1.json", 'r') as f:
    data = json.load(f)
    data = data["datasets"]["data-fed8b51d9e6940d5dce0ed1990c40e4f"]
    example = pd.DataFrame(data)
    
alt.Chart(example).mark_point().encode(
    x = "x",
    y = "y",
    color = "type",
    tooltip = ["x", "y"]
).interactive()

alt.Chart(...)

In [42]:
example = None
with open("file2.json", 'r') as f:
    data = json.load(f)
    data = data["datasets"]["data-2435179c396ce388c7701d192fb69c9a"]
    example = pd.DataFrame(data)
    
alt.Chart(example).mark_point().encode(
    x = "x",
    y = "y",
    color = "type",
    tooltip = ["x", "y"]
).interactive()

alt.Chart(...)

In [43]:
example = None

with open("iris1_example.json", 'r') as f:
    data = json.load(f)
    data = data["datasets"]["data-4cfc8fa242139f3561ce005edfd73548"]
    example = pd.DataFrame(data)
    
alt.Chart(example).mark_point().encode(
    x=x,
    y=y,
    color='species',
    tooltip=(x,y)
).interactive()

alt.Chart(...)

In [44]:
example = None

with open("iris2_example.json", 'r') as f:
    data = json.load(f)
    data = data["datasets"]["data-7a3bff88280d9682a8ee0f77d8298717"]
    example = pd.DataFrame(data)
    
alt.Chart(example).mark_point().encode(
    x=x,
    y=y,
    color='species',
    tooltip=(x,y)
).interactive()

alt.Chart(...)

In [45]:
example = None

with open("burtin1_example.json", 'r') as f:
    data = json.load(f)
    data = data["datasets"]["data-8833440f737304311e72aeeffd7b9881"]
    example = pd.DataFrame(data)
    
alt.Chart(example).mark_bar().encode(
    x = "Bacteria",
    y = alt.Y("Penicillin", scale=alt.Scale(type='pow', exponent=0.25)),
    color=alt.Color("Gram_Staining"),
    tooltip=[alt.Tooltip("Penicillin"), alt.Tooltip("Gram_Staining")]
)

alt.Chart(...)

In [46]:
example = None

with open("burtin2_example.json", 'r') as f:
    data = json.load(f)
    data = data["datasets"]["data-32830386a0d81a3f2e439bb9ba796d09"]
    example = pd.DataFrame(data)
    
alt.Chart(example).mark_bar().encode(
    x = "Bacteria",
    y = alt.Y("Penicillin", scale=alt.Scale(type='pow', exponent=0.25)),
    color=alt.Color("Gram_Staining"),
    tooltip=[alt.Tooltip("Penicillin"), alt.Tooltip("Gram_Staining")]
)

alt.Chart(...)

In [47]:
# source = vega_datasets.burtin()

# alt.Chart(source).mark_bar().encode(
#     x = "Bacteria",
#     y = alt.Y("Penicillin", scale=alt.Scale(type='pow', exponent=0.25)),
#     color=alt.Color("Gram_Staining"),
#     tooltip=[alt.Tooltip("Penicillin"), alt.Tooltip("Gram_Staining")]
# )

In [48]:
# # burtin10_source.json

# with open("../vis-dif/public/data/burtin15_source.json", 'r') as f:
#     data = json.load(f)
#     data = data["datasets"]["data-c6959e44dc4ee45f2be77dfe2deb6e78"]
#     example = pd.DataFrame(data)
    
# alt.Chart(example).mark_bar().encode(
#     x = "Bacteria",
#     y = alt.Y("Penicillin", scale=alt.Scale(type='pow', exponent=0.25)),
#     color=alt.Color("Gram_Staining"),
#     tooltip=[alt.Tooltip("Penicillin"), alt.Tooltip("Gram_Staining")]
# )